Refactor of the notebook endpoint

lucferbux opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->


## Description
<!--- Describe your changes in detail -->
Refactor of the `/notebooks` endpoint to secure the calls and avoid the injection of code in the notebook controller specification.

Fix #581 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

andrewballantyne commented on backend/src/routes/api/notebooks/index.ts:
```diff
@@ -61,39 +63,64 @@ module.exports = async (fastify: KubeFastifyInstance) => {
   );
 
-  fastify.patch(
+  fastify.delete(
```
No delete notebook endpoint please... this is not a needed feature.

lucferbux replied:
mmmm ok, what do you suggest, add a value for the PUT method and and if it's enable stop the notebook, or just a straight up endpoint to just stop. I would rather have the stop functionality separated somewhere else.

andrewballantyne replied:
I listed endpoints what I thought were good here based on what the UI needs to do -- https://github.com/opendatahub-io/odh-dashboard/issues/581

Thinking a bit more on it -- I think essentially we just need:
1. `GET` Notebook (fairly straight forward)
1. `GET` Notebook status (gets the notebook + its running status, which is more than just 
1. `POST` "Create" Notebook (which takes form data instead of a CR)
    - `PUT` "Restart" Notebook (which takes form data instead of a CR) -- honestly this could probably be part of the `POST` call and just check if we have a notebook, update/restart it, if we don't create it
1. `POST` "Stop" Notebook (takes nothing, applies the annotation server side)

4 endpoints might be enough to do all we need and nothing takes a CR anymore.

lucferbux replied:
Ok, seems great, sorry i get carried away and i forgot you had the spec, i'm gonna uodate the pr with this, it wont take much time, thanks @andrewballantyne

lucferbux replied:
Ok, updated with this endpoints you suggested, I've also updated the type `NotebookData`, planning to remove the following fields too:

* `imageSelection`& `imageUrl`: Just have an `image` field and then get the values in the backend, this will prevent any injection of undesired images.
*`envVarFileName` and just generate it in the backend, avoid sanitization for the name, add similar checks for the secrets and configmap values just to avoid the use of undesired values.

Apart from that, I've tested that we can move the PVC creation and the env creation to the same request of stop/restart, I wanna know your thoughts on that @cfchase @andrewballantyne.

cfchase replied:
Please do me a favor and don't put verbs like create/stop/start verbs in restful design.  It's an antipattern that absolutely drives me nuts.  the verbs should be the HTTP verbs.

https://learn.microsoft.com/en-us/azure/architecture/best-practices/api-design#organize-the-api-design-around-resources

lucferbux replied:
Ok fair, I followed this patter first but then I changed it, seems weird the stop status, I went first with `DELETE` since we were removing the resource of the notebook, I'm gonna put the update and stop inside the `PUT` and then the creation in `POST` to follow the specification. Thanks Chris.

cfchase replied:
a PUT is generally a full replacement of a resource.  PATCH for partial.

andrewballantyne replied:
POST for stop was my suggestion -- I guess I have seen a lot of endpoints in various stages of bastardization of REST... POST seems to be the "do action" in most APIs I have seen. PATCH was intended for providing some indication of what to adjust -- whereas here we would not ask anything of the client, a dedicated endpoint to do something `POST /api/notebooks/stop` seemed very reasonable to me. I of course defer to Chris, I'm an outsider when it comes to backend design.

cfchase replied:
POST has been the catchall abused verb since the beginning, but I'd rather not resort to that unless you really can't express the intent some better way.

andrewballantyne replied:
I agree Chris -- we probably shouldn't favour POST for the catchall situations and I'll bow to your judgement here. 

However PATCH seems wrong to me on the principle of "stop has nothing to do with PATCH" as far as our REST API is concerned. It's just how we talk to k8s to achieve the kubeflow desire of patching an annotation. Seems "leaking implementation" to me... not that doing that is a problem, just not one I agree with fully.

Anyways, I should stay out of backend design 🙂 PATCH for stop will work fine from an implementation point of view. 👍

lucferbux replied:
The main problem here is that following the full specification both of you are right:

* `PUT` should be idempotent and replace the hole object. (We are trying to avoid this since we don’t wanna mess around with the hole notebook spect).
* `PATCH` should use a patch language like `JSON` or `XML`, the body itself should contain the instructions to update the object but it doesn’t need to be idempotent.

Trying to hide away the spec of notebook makes both verbs harder to follow, so right now we can follow the specification we agree on (trying to avoid antipatterns like the one chris pointed out).

I’m fine either having stop as a `PUT` or `PATCH`, whatever you guys think it’s the best solution.

lucferbux replied:
If it’s a `PATCH`, we can add a sanitization in the backend to ensure the only JSON permitted is:

`metadata: { annotations: { 'kubeflow-resource-stopped': <date>} } `

With that, we ensure are following the `PATCH` spec and we are avoiding the modification of the rest of the notebook object.

What do you think?

cfchase replied:
I like it.

andrewballantyne replied:
Eliminate all you can from the frontend -- the frontend can talk "actions" not data... if we do that, then there can't be security holes, because they talk actions and we do all the data in a secure location.

Ideally, the call to stop a notebook requires NOTHING from the caller. It is a direct endpoint to say "do action" and thus however we need to do it can be done by us. If the caller needs to provide direction, let us make that as small as possible -- like if they need to provide a date, they JUST provide a date, we do the annotation and formatting of the date on the backend. 

The client can't be trusted, so make it hard for them to "lie" about what they are doing.

andrewballantyne replied:
The more the client gives us, the more sanitization we need to do... let us create as little of those points as possible to avoid any chance the UI can sneak something in.

lucferbux replied:
Yes, you are right too @andrewballantyne , I mean, to be fair, the one verb that fits the best action to stop for me is `DELETE` 😂. At the end of the day we are deleting the notebook, although we are just patching the spec in the backend. I agree adding PATCH adds more vulnerable surface. We are having a great discussion here, I would gladly accept both opinions.

andrewballantyne replied:
I retract any statement of verbs I have made -- just make it so the client only provides the extreme bare min -- easier for us to validate. Honestly, the backend can stop the Notebook at it's own timestamp if it wants to. Less chance the client can try and fudge the data.

cfchase replied:
we are not deleting the notebook resource

andrewballantyne replied:
@cfchase If you infer the date from the server and not the client -- yeah, I think stop is as simple as `PATCH /api/notebooks` with `{ state: 'stopped' }` and let the server apply the annotation & figure out the current datestamp to use as the stop.

lucferbux replied:
I'm implementing the suggestion you are discussing above as I think is the best way to do it too. Refactoring.

cfchase commented on backend/src/routes/api/notebooks/index.ts:
```diff
@@ -61,42 +63,59 @@ module.exports = async (fastify: KubeFastifyInstance) => {
 
   fastify.post(
-    '/:namespace',
+    '/create',
```
no need to create.  That's implied by the http verb.

cfchase commented on backend/src/routes/api/notebooks/index.ts:
```diff
@@ -61,42 +63,59 @@ module.exports = async (fastify: KubeFastifyInstance) => {
-  fastify.patch(
-    '/:namespace/:name',
+  fastify.put(
+    '/restart',
```
PATCH {state: `stop`}

andrewballantyne replied:
Starting needs data -- we have a stale Notebook, you need to be able to accept new configurations when we start it back up.

cfchase replied:
good point

cfchase commented on backend/src/routes/api/notebooks/index.ts:
```diff
@@ -61,42 +63,59 @@ module.exports = async (fastify: KubeFastifyInstance) => {
     ),
   );
+
+  fastify.post(
```
same endpoint as start `{"state": "stop"}

cfchase commented on backend/src/routes/api/notebooks/index.ts:
```diff
@@ -61,39 +63,78 @@ module.exports = async (fastify: KubeFastifyInstance) => {
+    ),
+  );
+
+  fastify.put(
```
>    * `PUT` "Restart" Notebook (which takes form data instead of a CR) -- honestly this could probably be part of the `POST` call and just check if we have a notebook, update/restart it, if we don't create it

Now that @andrewballantyne has reminded me that starting the notebook takes the same data as the create, I think he's right.  Can we just get rid of the PUT altogether and do a get at the beginning of the post and act accordingly?

lucferbux replied:
sounds right to me, having that kind of logic in the post seems right, I would agree, what do you think @andrewballantyne ?

andrewballantyne replied:
Yup, that seems the most logical... it'll be doing a lot of the same code for create or update & start... so it definitely seems wise to make 1 endpoint to avoid having to figure out how to share 90% of the code between the two flows.

lucferbux commented:
UPDATE: I've changed the endpoints, just delete the PVC creation in the frontend and refactored some parts of the notebook logic, my goals left for this PR are:

1. Refactor the environment variables creation and delete the configmap and secret routes.
2. Add sanetization just to the NotebookData values.

I'm aiming to have that for today or tomorrow.

andrewballantyne commented:
> `getting-started` endpoint is missing (removed by some other PR, don't have the link). Is it expected? if not, could you add it back in this PR?

For bookkeeping and sanity reasons, please don't make unrelated requests on our PRs @bdattoma -- for this issue though, the ISV removed it because they are not CRs.

andrewballantyne commented on manifests/base/cluster-role.yaml:
```diff
@@ -83,6 +83,14 @@ rules:
+    apiGroups:
+      - user.openshift.io
+    resources:
+      - users
```
Not sure we will need this unless I get a clear answer on the `username` front that makes sense.

lucferbux replied:
replied above

andrewballantyne commented:
From my last review -- just to help with importance.

Issues I want clarified:
- https://github.com/opendatahub-io/odh-dashboard/pull/587#discussion_r990291620
    - https://github.com/opendatahub-io/odh-dashboard/pull/587#discussion_r990294400
- https://github.com/opendatahub-io/odh-dashboard/pull/587#discussion_r990304072

I think this will be a legit error that Security won't like:
- https://github.com/opendatahub-io/odh-dashboard/pull/587#discussion_r990307885

lucferbux commented on backend/src/routes/api/notebooks/index.ts:
```diff
@@ -1,23 +1,26 @@
-module.exports = async (fastify: KubeFastifyInstance) => {
+export default async (fastify: KubeFastifyInstance): Promise<void> => {
   fastify.get(
     '/:namespace/:name',
```
I think I added it cause it was raising a linter error, but I cannot recall right now, I can test it to see if that was the case.

lucferbux replied:
Yup, it triggers a `Missing return type on function  @typescript-eslint/explicit-module-boundary-types`

andrewballantyne replied:
That's a confusing lint error -- I'll have to investigate this -- this has nothing to do with a return type 😕❓

lucferbux commented on backend/src/routes/api/notebooks/utils.ts:
```diff
@@ -0,0 +1,129 @@
+  return pods.some((pod) => checkPodContainersReady(pod));
+};
+
+export const checkPodContainersReady = (pod: V1Pod): boolean => {
```
mmm sure, i might have slipped this export.

lucferbux commented on backend/src/utils/notebookUtils.ts:
```diff
@@ -0,0 +1,620 @@
+import { getDashboardConfig } from './resourceUtils';
```
Yup, i agree, we have too much stuff gathered into a generic utils, might be split up.

lucferbux commented on backend/src/typeHelpers.ts:
```diff
@@ -1,8 +1,73 @@
+ * TODO: Implement the SDK & Patch logic -- this should stop being needed as things will be defined as Patches
  */
-export type RecursivePartial<T> = {
+ export type RecursivePartial<T> = {
```
Ok, let me check it

lucferbux commented on backend/src/utils/route-security.ts:
```diff
@@ -70,6 +44,28 @@ const testAdmin = async (
   return false; // Not admin, no bypassing
 };
 
+const requestSecurityGuardNotebook = async (
```
No, this is for other purposes, the username field is for the ability to spawn other users' notebooks. If not present, it will get the username by the request. If added, admins can spanw notebooks for existing users and regular users will have a returning error.

lucferbux commented on backend/src/types.ts:
```diff
@@ -666,3 +659,62 @@ type GroupCustomObjectItemMetadata = {
+  gpus: number;
+  envVars: EnvVarReducedTypeKeyValues;
+  state?: NotebookState;
+  username?: string;
```
Yes, we get the username from oauth, this is to add the feature to allow admin spawn other users' notebooks

lucferbux commented on frontend/src/pages/notebookController/const.ts:
```diff
@@ -2,7 +2,6 @@ import { NotebookControllerUserState } from '../../types';
 
 export const CUSTOM_VARIABLE = 'Custom variable';
 export const EMPTY_KEY = '---NO KEY---';
-export const MOUNT_PATH = '/opt/app-root/src';
```
No longer used, MOUNT_PATH is used for the notebook creation and moved to there, we can leave it there, but it's not used in the rest of the project.

lucferbux commented on backend/src/types.ts:
```diff
@@ -666,3 +659,62 @@ type GroupCustomObjectItemMetadata = {
+  notebookSizeName: string;
+  imageName: string;
+  imageTagName: string;
+  url: string;
```
ok, i sure missed that, can change that later

lucferbux commented on backend/src/utils/route-security.ts:
```diff
@@ -175,6 +171,8 @@ const handleSecurityOnRouteData = async (
     );
+  } else if (isRequestNotebookAdmin(request)) {
+    await requestSecurityGuardNotebook(fastify, request, request.body.username);
   } else {
```
only the ones that contains the username attribute right? if not that's an issue, but I've tested this with regular users.

cfchase replied:
doing a normal notebook creation:
```
{"level":40,"time":1665170114051,"pid":68,"hostname":"rhods-dashboard-86bf7c6b85-pqzf6","msg":"User user1 interacted with a resource that was not secure."}
```

lucferbux replied:
Yes, in the last change i messed up that logic and it get's to that, thanks for the check!

andrewballantyne commented on backend/src/types.ts:
```diff
@@ -666,3 +659,62 @@ type GroupCustomObjectItemMetadata = {
+  url: string;
+  gpus: number;
+  envVars: EnvVarReducedTypeKeyValues;
+  state?: NotebookState;
```
```suggestion
  state: NotebookState;
```
Couldn't find a situation where this being optional was helpful/used. Was this just a transition in the code as you developed?

lucferbux replied:
oh, i was doing the refactor with all the improvements and I ended up adding that same thing. I made it optional when we were discussing the way to implement the stop and start, but it ended up being mandatory. tx!

andrewballantyne reviewed (commented):
#643 will follow up this PR. I made `state` an expected part of the notebook endpoints that are directly `/api/notebooks`

lucferbux commented on backend/src/utils/route-security.ts:
```diff
@@ -70,6 +44,28 @@ const testAdmin = async (
+  await testAdmin(fastify, request, true);
+
+  try {
+    await getOpenshiftUser(fastify.kube.customObjectsApi, username);
```
👌

lucferbux commented on backend/src/utils/route-security.ts:
```diff
@@ -70,6 +44,28 @@ const testAdmin = async (
+  request: OauthFastifyRequest,
+  username: string,
+): Promise<void> => {
+  const user = await getUserName(fastify, request);
```
yup, i think it was due to the last refactor I did fixing some QA's requirements, I sure rushed things out.

lucferbux commented:
Closing this PR in favor of #587
